Minor changes to `StructuredTaskScope`

PaulSandoz opened the pull request:
Some suggested minor changes to `StructuredTaskScope` source while reviewing the code.

Overall I like how this approach has boiled down to about as simple as it can get while supporting the constraints of structured concurrency (thanks in no small part to `ThreadFlock`).

I think it should be possible to simplify the subclass implementations if the values of `Future.State` were comparable as in `RUNNING` < `CANCELED` < `FAILED` < `SUCCESS`. Otherwise, introducing that internally could also simplify and avoid the task scopes holding on unnecessarily to futures that are not operated on after join. I was reluctant to make such a change here but could propose as a separate PR.

I was uncertain if the `ShutdownOnSuccess.result` methods and the various exception returning/operating methods on `ShutdownOnFailure` should be constrained to throw if called before `join` has completed. Otherwise, we should specify that if operated on before `join` has completed then the results are unspecified.
(There might be a simple way for join to returning something other than `this`, thereby avoiding this issue, with some additional complexity, but I suspect you have already thought of that and took the minimal route.)

I presume there is an implicit happens-before edge on `join` that we can make explicit and specify?

pron commented:
Thank you!
We've explored policies different `join` signatures wrapping the STS class rather than extending it. Ultimately, we favoured a common supertype with a common `join` method that better exposes the similarities between them, and `join` returning `this` is a reasonable compromise between expressing the similarity and a convenient, typesafe API.

AlanBateman commented on src/jdk.incubator.concurrent/share/classes/jdk/incubator/concurrent/StructuredTaskScope.java:
```diff
@@ -269,17 +269,21 @@
 
-    // the set of "tracked" Future objects, created lazily
+    // the set of "tracked" Future objects waiting to be returned by Future.get, created lazily
+    // assigned once-only with FUTURES.compareAndSet, read by any thread
```
Instead of "assigned once-only ..." it might be simpler to just link to the track method

PaulSandoz replied:
Changed to "    // assigned to non-null value in method track, read by any thread"

AlanBateman commented on src/jdk.incubator.concurrent/share/classes/jdk/incubator/concurrent/StructuredTaskScope.java:
```diff
@@ -342,6 +346,7 @@ private boolean isShutdown() {
      */
     private void track(Future<?> future) {
         // create the set of Futures if not already created
+        // cannot use double-checked locking since it requires use of a synchronized block
```
I'd prefer to leave this comment as is, only because mentioning another possible implementation might confuse reader.

PaulSandoz replied:
Ok, removed. When/if monitors are supported properly in loom we can use double checked locking, which is more idiomatic. Hard to track these things.

AlanBateman reviewed (commented):
Sub-class -> subclass, shutdown -> shut down, and moving the unchecked cast to FutureImpl are good - thanks. Just a few comments on the changes to the internal comments.

PaulSandoz commented:
I added a class doc section on memory consistency effects, and updated the access methods on the subclasses. 

I propose another PR, once this one has integrated, on the `ShutdownOnSuccess` and `ShutdownOnFailure` implementations to contrast and lets see if its useful.

PaulSandoz commented on src/jdk.incubator.concurrent/share/classes/jdk/incubator/concurrent/StructuredTaskScope.java:
```diff
@@ -251,6 +251,14 @@
  *
+ * <h2>Memory consistency effects</h2>
+ * <p>Actions in the owner thread of, or a thread contained in, the task scope prior to
+ * {@linkplain #fork forking} of a {@code Runnable} or {@code Callable} task
```
Doh! thanks. I also fixed the link.
